Merge drush audit tasks

steveworley opened the pull request:
This one adds a views pagination check - it will attempt to locate all view displays and will ensure that the pagination is not configured to be above a certain threshold. 

Pagination > 30 is seen as cause for concern as we don't have full scope of what the view is doing and can lead to OOM errors.

I also ported Module Settings check from my Drush Audit tool. This will let us define a set of modules and their recommended settings. One of the go live tasks we have (especially for govCMS projects) is that we ensure Login Security is enabled and configured.

steveworley committed:
Add views pagination check

steveworley committed:
Adds views pagination and module settings checks

- Adds an iterator base class for iterator through settings
- Adds a views check for pagination

steveworley committed:
Update views pagination and module settings check

- Update views pagiantion check to use Seriliazer to prevent unserialization errors
- Updates govcms_saas configuration so it passes TRUE = 1 and FALSE = 0 as expected
- Updates settings iterator so that it properly iterates via ::next()

seanhamlin commented:
I have long been thinking about having a generic variable getting and checking script (similar to module enabled etc). This PR looks amazing, will review in more detail later today. Thanks!

steveworley commented:
Awesome! Let me know if there's anything you want changed.

I've done it in a way that you can implement a SettingsIterator to check different type of settings (like if your module manages settings in its own table). Hopefully that gives enough flexibility over how to define settings:

``` yml
settings:
    \Iterator\Class\NameSpace:
         setting_name: [expected value]
```

One caveat is checking truthy/falsey settings. Variables are translated to 0 and 1 which may get confusing - so I think commenting around the settings in the profiles might alleviate some of this. [This commit](https://github.com/seanhamlin/site-audit/pull/51/commits/f181864d87fcc847b8b75a0460bb046a5bcb9aec#diff-95f5619df679d93bbc156a983c7bb524) is where I updated the settings.

seanhamlin commented on src/Settings/VariableIterator.php:
```diff
@@ -0,0 +1,22 @@
+   */
+  public function valid() {
+    $actual = $this->drush()->getVariable($this->key());
+    return $this->current() == $actual;
```
A double equals will help with the `TRUE` vs `1` etc that may be in the settings. It might pay to add a comment to explain this. I generally use triple equals in PHP code as a rule of thumb.

steveworley replied:
Yeah- that is exactly why I used double equals truthy checks were an issue here.

seanhamlin commented on src/Settings/VariableIterator.php:
```diff
@@ -0,0 +1,22 @@
+   * @return bool
+   */
+  public function valid() {
+    $actual = $this->drush()->getVariable($this->key());
```
No default value here, should that be `NULL` or similar to be explicit?

seanhamlin commented on profiles/govcms_saas.yml:
```diff
@@ -88,3 +88,23 @@ checks :
+    machine_name: 'login_security'
+    enabled: true
+    settings:
+      \SiteAudit\Settings\VariableIterator:
```
Is it useful to have pluggable iterators? I think this would make the most sense if this just was the iterator that all moduleSettings checks use. Thoughts?

steveworley replied:
Plugable iterators allow us to define a different way to check settings if we need to. Say; a module maintains a configuration table separately- we could define another iterator to query DBs as opposed to variable checks.

seanhamlin commented on profiles/govcms_saas.yml:
```diff
@@ -88,3 +88,23 @@ checks :
+    threshold: 100
+
+  # Module configuration checks.
+  \SiteAudit\Check\D7\ModuleSettings:
```
Is it possible to have multiple `ModuleSettings` checks declared in the same profile?

steveworley replied:
Good point- I didn't check that. Will test.

seanhamlin commented on src/Check/D7/ViewsPagination.php:
```diff
@@ -0,0 +1,63 @@
+      }
+
+      if ($display_options['pager']['options']['items_per_page'] > $this->getOption('threshold', 30)) {
+        $errors[] = "<b>$view_name <i>[$display_name]</i></b> is displaying <code>{$display_options['pager']['options']['items_per_page']}</code>";
```
use `<strong>` and `<code>`, not `<b>` and `<i>`

steveworley committed:
Module settings checks

- Updates profiles so they support a `settings` key

steveworley commented:
Hey @seanhamlin I think I've addressed the review. Settings check results are bundled with the other check results. So I added translations to all `AuditReponse` methods so I could output the name of the module at each step otherwise you might end up with a bunch of *Module settings* checks which probably aren't very helpful.

![image](https://cloud.githubusercontent.com/assets/1840912/22284531/7a6e8858-e333-11e6-8b5d-38e4a6a873f9.png)

We might be able to split these up as well so we have "checks" and "configuration" or similar.

seanhamlin commented on profiles/govcms_saas.yml:
```diff
@@ -88,3 +88,23 @@ checks :
   \SiteAudit\Check\Phantomas\InPage404s : {}
+
+  # Field specific checks.
+  \SiteAudit\Check\D7\EntityReferenceAutocomplete:
```
I moved this up the YAML file, so this is now in there twice.

seanhamlin commented on src/AuditResponse/AuditResponse.php:
```diff
@@ -49,6 +60,31 @@ public function __toString() {
+   * @return string
+   *   The translated string.
+   */
+  private function translate($message, $tokens = []) {
```
Nice addition!

seanhamlin commented:
Will fix up the profile in a follow up commit. Thanks @steveworley !
